gensim/utils.py

```diff
@@ -139,6 +140,36 @@ def file_or_filename(input):
         return input
 
 
+@contextmanager
+def open_file(input):
+    """ Generates 'with' like behaviour except closing the file object
+
+    Parameters
+    ----------
+    input : str or file-like
+        Filename or file-like object.
+
+    Yields
+    -------
+    str
+        lines from the given file-like object or a new file-like object from the given filename.
+
+    """
+    mgr = file_or_filename(input)
+    exc = False
+    try:
+        yield mgr
+    except Exception:
+        # Handling any unhandled exceptions from the code nested in 'with' statement.
+        exc = True
+        if not type(mgr).__exit__(mgr, *sys.exc_info()):
```

Comment:
why `type` is needed here?

Reply:
As per the so thread , " exit = type(mgr).__exit__ is used because CPython has a history of caching special methods and this just makes it consistent ".